File: src/test/java/io/reactivesocket/cli/i9n/TestOutputHandler.java
package io.reactivesocket.cli.i9n;

import com.google.common.collect.Lists;
import io.reactivesocket.cli.OutputHandler;
import io.reactivesocket.cli.UsageException;

import java.util.List;

public final class TestOutputHandler implements OutputHandler {
    public final List<String> stdout = Lists.newArrayList();
    public final List<String> stderr = Lists.newArrayList();

    @Override
    public void showOutput(String s) {
        stdout.add(s);
    }

    @Override
    public void info(String s) {
        stderr.add(s);
    }

    @Override
    public void error(String msg, Throwable e) {
        if (e instanceof UsageException) {
            stderr.add(e.getMessage());
        } else {
            stderr.add(msg + ": " + e.toString());
        }
    }

    @Override
    public int hashCode() {
        return stdout.hashCode() + stderr.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof TestOutputHandler)) {
            return false;
        }

        TestOutputHandler other = (TestOutputHandler) obj;

        return stderr.equals(other.stderr) && stdout.equals(other.stdout);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder(4096);

        if (!stdout.isEmpty()) {
            sb.append("STDOUT:\n");
            stdout.forEach(s -> sb.append(s + "\n"));
        }

        if (!stderr.isEmpty()) {
            sb.append("STDERR:\n");
            stderr.forEach(s -> sb.append(s + "\n"));
        }

        return sb.toString();
    }
}

File: src/test/kotlin/io/rsocket/cli/i9n/TestOutputHandler.kt
package io.rsocket.cli.i9n

import com.baulsupp.oksocial.output.OutputHandler
import com.baulsupp.oksocial.output.UsageException
import com.google.common.collect.Lists

class TestOutputHandler : OutputHandler<Any> {
  private val stdout: MutableList<String> = Lists.newArrayList()
  private val stderr: MutableList<String> = Lists.newArrayList()

  override suspend fun showOutput(output: Any) {
    stdout.add(output.toString())
  }

  override fun info(msg: String) {
    stderr.add(msg)
  }

  override suspend fun showError(msg: String?, e: Throwable?) {
    if (e is UsageException) {
      stderr.add(e.message.toString())
    } else {
      stderr.add(msg + ": " + e.toString())
    }
  }

  override fun hashCode(): Int {
    return stdout.hashCode() + stderr.hashCode()
  }

  override fun equals(other: Any?): Boolean {
    if (other is TestOutputHandler) {
      return stderr == other.stderr && stdout == other.stdout
    }

    return false
  }

  override fun toString(): String {
    val sb = StringBuilder(4096)

    if (!stdout.isEmpty()) {
      sb.append("STDOUT:\n")
      stdout.forEach { s -> sb.append(s).append("\n") }
    }

    if (!stderr.isEmpty()) {
      sb.append("STDERR:\n")
      stderr.forEach { s -> sb.append(s).append("\n") }
    }

    return sb.toString()
  }
}
